### ISSUE TITLE
Server crashes after working with Mint

### ISSUE THREAD

**Issue by goertzenator** [StrangeLoopGames/EcoIssues#2721]
## Summary
1. A Mint was placed into the world and exhibited some weirdness.
2. Existing clients soon crashed out.
3. Restarting the server results in a crash shortly after startup.  The world is 30 days old and now unusable.

### Mint Placement
1. I placed a Mint that I had built a few days prior.
2. I started playing around with the currency name.  I named it "sdfsdf", and then hit `currency report` for a very uninteresting report.
3. I then changed the currency name to just "s" and hit `currency report` again.  It provided a report for the previous "sdfsdf" instead of "s".
4. I then changed the currency name "xxx" and hit `currency report`.  It correctly provided a report for "xxx".
5. Around the same time I placed a Currency Exchange but didn't do anything with it.
6. I left the game moments later and went to bed.

Perhaps this has nothing to do with the bad things that followed, but I mention it because it was novel for this server.

### No Client Logins
Eco user Dmorchard (github @dmorchard) reported the following to me about 5 minutes later...
```
Caught exception: Caught exception invoking RPC ViewRPC on Client!

Caught exception invoking RPC PerformTrade on StoreComponent!

Value cannot be null.
Parameter name: key


Stack:
   at System.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Collections.Generic.Dictionary2.FindEntry(TKey key)
   at Eco.Gameplay.Items.InventoryChangeSet.MoveItems[TKey](Type itemType, Int32 quantity, Inventory source, Inventory destination, Func2 sourceStackPredicate, Func`2 sourceStackOrder)
   at Eco.Gameplay.Components.StoreComponent.PerformTrade(Player purchaser, BSONObject tradeData)



Stack:
   at Eco.Shared.Networking.RPCManager.TryInvoke(Object controller, String methodname, BSONObject bsonArgs, Object& result)
   at Eco.Shared.Networking.RPCManager.InvokeOn(BSONObject bson, Object controller, String methodname)
   at Eco.Plugins.Networking.Clients.Client.ViewRPC(Guid guid, String methodname, BSONObject bson)
```
Now I can log in but can't do anything.  Then it kicks me for "inactivity". 

### Server Crash

The server starts and runs for a few seconds before crashing:

![image](https://user-images.githubusercontent.com/605072/32357275-546d8f2e-c009-11e7-98d8-981d4ed9dda1.png)

Download the  [crashing world](https://www.dropbox.com/s/qi1k2w80naloyx1/server.zip?dl=0).

**Comment by dmorchard** [comment]
I was attempting to sell goods to goertzenator's store (had the GUI open) at the very moment he was creating a new currency - I remember seeing the notifications appear in my chat window.  I can't recall precisely what followed, but I wasn't able to complete the transaction and shortly my client exited the server with the error message posted above.  I made a few attempts to log back in which appeared to work at first, but once in-game I was unable to interact with anything, and after a short period I'd get dumped off the server with an error message that said simply, "inactivity".  After 2 or 3 attempts with the same result I gave up.

**Comment by MarjieAVolk** [comment]
Server crash could be related to https://github.com/StrangeLoopGames/EcoIssues/issues/2642?

Store crash looks unrelated - unless maybe the trade you tried to perform modified a stockpile?

**Comment by dmorchard** [comment]
My trade would've added ore (or maybe it was coal) to goertzenator's stockpile, yes.  I had successfully traded several cart loads at his store in the time leading up to the crash.

**Comment by MarjieAVolk** [comment]
Is there a chance one of the trade offers in the trade had no item selected?

**Comment by dmorchard** [comment]
There's a chance - it's very tough to recall what happened b/c it was nearly 48 hours ago now :)  In fact, the way the store GUI works I would think that a zero-item trade would be a fairly simple mistake to make (for valid trades a 1st click on the button zeroes out the items, and a 2nd accidental click would result in a zero-item trade).  I'd expect that to happen from time to time.  So are you hypothesizing it was the a zero-item trade in combination with (or following) the mint interactions, specifically?

**Comment by dmorchard** [comment]
What about our hopes of salvaging the world?  It seems that the server continued to run for several hours after I was ejected - a third server member reports playing without noticing anything very out of the ordinary - just trouble getting his own research table connected to his own stockpile, which I believe he resolved.  If I understand how the backups work this means we have no uncorrupted backup file.

**Comment by MarjieAVolk** [comment]
To clarify, I didn't mean a trade with zero items, I meant a trade that attempted to trade an invalid listing in the store - like the store had a listing with no item selected there, just the plus icon showing.

As for salvaging the world, if it's still crashing with that same exception, it's possible you could delete stockpiles out of the save file until you find the bad one.  Cheat back in the stuff you lost, etc.  You would need to open up the save file like a .zip, open up the WorldObject folder and delete files with 'Stockpile' in their name.

**Comment by johnkslg** [comment]
Dont suppose you folks have this failing server save handy somewhere?

**Comment by goertzenator** [comment]
Deleting Stockpile-48a3fd allowed the world to start.   Apologies for the delayed response; we were trying to find the in-game stockpile that this file represented but couldn't.  Link to crashing world at bottom of original bug report.

**Comment by johnkslg** [comment]
Migrated the server and ran it, works fine now.  Also spent time experimenting with stockpiles, currencies, changing currency names, stores with full inventories, etc.  Couldnt get any of the stuff you described to occur. Going to close for now, I think it was fixed by some other bug fixes out there.   Let me know if it crops up again. 

I did fix mints so that setting them to an invalid name will reset the text.
